mitodl/mit-open
# Setup codespaces

### What are the relevant tickets?
Closes https://github.com/mitodl/mit-open/issues/1382

### Description (What does it do?)
This PR adds a Github Codespace configuration do  we can setup disposable live and shareable demo environments right from github. The idea is that we would will now have shareable environments for PRs which we can share with the design/product team for feedback and also be able to test any feature that requires a publicly accessible endpoint.

### How can this be tested?
The codespace configuration for projects is sourced from the main branch so in order to put this together (and test) I had to create a personal fork of mit-open with these changes on main.

1. On the top right of this PR click the green "code" button and go to the codespaces tab and launch a codespace
<img width="927" alt="Screenshot 2024-08-20 at 3 51 02 PM" src="https://github.com/user-attachments/assets/d27aba57-fafc-4380-b5c7-573afa4ccb87">
2. This should start provisioning a code space in a new tab (this takes a bit of time so grab a coffee). You will eventually see some output that indicates the script is waiting for migrations to finish before fetching some data
<img width="1046" alt="Screenshot 2024-08-20 at 4 08 53 PM" src="https://github.com/user-attachments/assets/cadb2362-9a5a-43b1-bad1-d8d84b28e3e8">
3. Once this completes, the codespace should automatically load up the readme and the "ports" tab will indicate that there are 3 ports exposed.
4. There is a know issue in codespaces with CORS unless the backend endpoint is set to public. To do this, go to the ports tab and set port "8063" to public. For this reason also - DEBUG is set to False for codespaces in django 
<img width="1145" alt="Screenshot 2024-08-20 at 4 15 51 PM" src="https://github.com/user-attachments/assets/4f5a8768-0f15-41d6-8abe-ad37120468e6">
5. You should at this point be able to visit the port 8602 endpoint and load it up. If it doesn't load -  try again in 2 minutes. it is likely that the frontend is still building.
6. If you want data to be automatically fetched upon loading up the codespace, you can add relevant secrets such as the edx client secret and key to your codespace secrets under the [codespace config page](https://github.com/settings/codespaces)


### Additional Context
- There are integrations with jetbrains and [visual studio](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-github-codespaces-in-visual-studio-code) which might be worth trying out
- Codespaces are provisioned under your own personal account - each developer gets 120 free hours every month which I think should be more than enough considering not every PR will require a codespace. Your persoinal github settings page has some options for default instance sizes etc (i set mine to the smallest instance sizes).
- You can add secrets and more environment variables on the [codespace settings](https://github.com/settings/codespaces) page
- You will have to manually set the api's port to public each time you launch a codespace (noted in instructions above) due to a known limitation with cors in codespaces. see:
  - https://github.com/orgs/community/discussions/22399
  - https://github.com/orgs/community/discussions/31316
  - https://github.com/orgs/community/discussions/15351
